File: tools/testing/tui_test.py
#!/usr/bin/env python3
"""M4KK1 4P1 - tui_test.py
ncurses dashboard for `make test` (tools/testing/test_all.sh).

Runs the test script unchanged as a subprocess and parses its
output live:
  "=== 测试N: <title> ==="  -> section
  "[PASS] <msg>"            -> green pass entry (count)
  "[FAIL] <msg>"            -> red fail entry (count + list)
  "[INFO] <msg>"            -> info line (log only)
  "通过: N" / "失败: N"      -> script's own tally (cross-check)

Hotkeys: f follow toggle · s save report · PgUp/PgDn/↑↓ scroll ·
q quit.  Exit code mirrors the script's (0 all pass).
"""
import curses, os, re, selectors, subprocess, sys, time

ROOT = os.path.normpath(os.path.join(
    os.path.dirname(os.path.abspath(__file__)), "..", ".."))
LOGDIR = os.path.join(ROOT, "logs")

RE_SEC = re.compile(r"^===\s*(.+?)\s*===$")
RE_PASS = re.compile(r"^\[PASS\]\s*(.*)$")
RE_FAIL = re.compile(r"^\[FAIL\]\s*(.*)$")
RE_INFO = re.compile(r"^\[INFO\]\s*(.*)$")
RE_TALLY = re.compile(r"^通过:\s*(\d+)")
RE_TALLY_F = re.compile(r"^失败:\s*(\d+)")


class Log:
    def __init__(self):
        self.lines = []
        self.follow = True
        self.view = 0

    def push(self, raw):
        txt = re.sub(r"\x1b\[[0-9;]*[A-Za-z]", "",
                     raw.rstrip("\r\n"))
        self.lines.append(txt)
        if len(self.lines) > 20000:
            self.lines = self.lines[-10000:]
        if self.follow:
            self.view = max(0, len(self.lines) - 1)

    def scroll(self, dy, ph):
        self.follow = False
        self.view = max(0, min(len(self.lines) - 1,
                               self.view + dy))

    def page(self, dp, ph):
        self.scroll(dp * max(1, ph - 1), ph)


def stage_of(line):
    """live counters are kept by the caller; this returns the new
    section title when the line opens one, else None."""
    m = RE_SEC.match(line)
    return m.group(1) if m else None


def draw(scr, log, section, npass, nfail, rc, elapsed,
         running, saved):
    scr.erase()
    my, mx = scr.getmaxyx()
    ph = my - 3

    # header: title + live tally
    title = " M4KK1 make test — %s " % (section or "starting")
    try:
        scr.addstr(0, 0, title[:mx - 1].ljust(mx - 1),
                   curses.A_BOLD | curses.color_pair(2))
    except curses.error:
        pass
    tally = "PASS %d  FAIL %d  %s %5.1fs" % (
        npass, nfail,
        "running" if running else
        ("done rc=%s" % rc if rc is not None else "killed"),
        elapsed)
    try:
        scr.addstr(0, max(0, mx - 1 - len(tally)), tally[:mx - 1],
                   curses.color_pair(
                       4 if not running and nfail == 0 else
                       (3 if nfail else 4)))
    except curses.error:
        pass

    # log pane (pass/fail lines highlighted)
    top = log.view if not log.follow else max(
        0, len(log.lines) - ph)
    for i in range(ph):
        idx = top + i
        if idx >= len(log.lines):
            break
        l = log.lines[idx]
        attr = (curses.color_pair(4) if RE_PASS.match(l)
                else curses.A_BOLD | curses.color_pair(3)
                if RE_FAIL.match(l) else curses.A_NORMAL)
        try:
            scr.addstr(1 + i, 0, l[:mx - 1], attr)
        except curses.error:
            pass

    # footer
    foot = ("[f]ollow=%s  [s]ave report  "
            "[PgUp/PgDn] scroll  [q]uit %s"
            % (log.follow and "on" or "off", saved))
    try:
        scr.addstr(my - 1, 0, foot[:mx - 1].ljust(mx - 1),
                   curses.A_REVERSE)
    except curses.error:
        pass
    scr.timeout(100)


def save_report(log, npass, nfail):
    os.makedirs(LOGDIR, exist_ok=True)
    path = os.path.join(
        LOGDIR, "tui_test_%s.log"
        % time.strftime("%Y%m%d_%H%M%S"))
    with open(path, "w") as f:
        f.write("# make test report %s  PASS=%d FAIL=%d\n"
                % (time.strftime("%F %T"), npass, nfail))
        f.write("\n".join(log.lines))
    return "→ %s" % os.path.basename(path)


def main(scr, argv):
    curses.curs_set(0)
    curses.start_color()
    curses.use_default_colors()
    curses.init_pair(1, curses.COLOR_WHITE, -1)
    curses.init_pair(2, curses.COLOR_CYAN, -1)
    curses.init_pair(3, curses.COLOR_RED, -1)
    curses.init_pair(4, curses.COLOR_GREEN, -1)

    script = os.path.join(ROOT, "tools", "testing",
                          "test_all.sh")
    cmd = ["bash", script]
    proc = subprocess.Popen(
        cmd, cwd=ROOT, stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT, text=True, bufsize=1)

    log = Log()
    section = None
    npass = nfail = 0
    script_pass = script_fail = None
    saved = ""
    rc = None
    t0 = time.time()

    sel = selectors.DefaultSelector()
    sel.register(proc.stdout, selectors.EVENT_READ)

    scr.timeout(100)
    while True:
        drained = 0
        while drained < 500:
            events = sel.select(timeout=0)
            if not events:
                break
            for key, _ in events:
                line = key.fileobj.readline()
                if not line:
                    break
                log.push(line)
                # parse on the STRIPPED line: the script wraps
                # [PASS]/[FAIL] in ANSI colour codes, so the raw
                # line never matches the regexes
                clean = log.lines[-1]
                s = stage_of(clean)
                if s:
                    section = s
                if RE_PASS.match(clean):
                    npass += 1
                elif RE_FAIL.match(clean):
                    nfail += 1
                m = RE_TALLY.match(clean)
                if m:
                    script_pass = int(m.group(1))
                m = RE_TALLY_F.match(clean)
                if m:
                    script_fail = int(m.group(1))
                drained += 1
        if proc.poll() is not None:
            for line in proc.stdout:
                log.push(line)
                clean = log.lines[-1]
                if RE_PASS.match(clean):
                    npass += 1
                elif RE_FAIL.match(clean):
                    nfail += 1
            rc = proc.returncode
            break

        draw(scr, log, section, npass, nfail, rc,
             time.time() - t0, True, saved)
        k = scr.getch()
        ph = scr.getmaxyx()[0] - 3
        if k in (ord('q'), 27):
            if proc.poll() is None:
                proc.terminate()
            return 130
        elif k == ord('f'):
            log.follow = not log.follow
        elif k == ord('s'):
            saved = save_report(log, npass, nfail)
        elif k == curses.KEY_PPAGE:
            log.page(-1, ph)
        elif k == curses.KEY_NPAGE:
            log.page(1, ph)
        elif k == curses.KEY_UP:
            log.scroll(-1, ph)
        elif k == curses.KEY_DOWN:
            log.scroll(1, ph)

    # finished: show tail until user leaves
    while True:
        draw(scr, log, section, npass, nfail, rc,
             time.time() - t0, False, saved)
        k = scr.getch()
        ph = scr.getmaxyx()[0] - 3
        if k in (ord('q'), 27):
            break
        elif k == ord('s'):
            saved = save_report(log, npass, nfail)
        elif k == curses.KEY_PPAGE:
            log.page(-1, ph)
        elif k == curses.KEY_NPAGE:
            log.page(1, ph)
        elif k == curses.KEY_UP:
            log.scroll(-1, ph)
        elif k == curses.KEY_DOWN:
            log.scroll(1, ph)

    # cross-check our tally with the script's own summary
    if (script_pass is not None
            and script_pass != npass):
        log.lines.append(
            "tui_test: tally mismatch — parsed PASS %d "
            "vs script %s" % (npass, script_pass))
    return rc if rc is not None else 1


def run():
    os.environ.setdefault("ESCDELAY", "25")
    try:
        return curses.wrapper(
            lambda s: main(s, sys.argv[1:]))
    except RuntimeError as e:
        print("tui_test: %s (need >= 80x24 terminal)" % e,
              file=sys.stderr)
        return 2


if __name__ == "__main__":
    sys.exit(run())

File: tools/testing/tui_test.c
/*
 * M4KK1 4P1 - tui_test.c
 * ncurses test dashboard for tools/testing/test_all.sh (host tool,
 * WSL gcc + libncurses).  C port of tui_test.py.
 *
 * Runs the test script unchanged as a child process and parses its
 * live output (ANSI-colour stripped before matching):
 *   "=== 测试N: ... ==="  → section header
 *   "[PASS] ..." / "[FAIL] ..." → tally
 *   "通过: N" / "失败: N"      → script's own final tally (cross-check)
 *
 * UI: header (running/rc, PASS/FAIL counters, section), scrollable
 * log pane (PASS green / FAIL red), footer hotkeys: f follow, s save
 * report to logs/tui_test_<ts>.log, PgUp/PgDn/UP/DOWN scroll, q quit.
 * Exit code = child exit code.
 *
 * Copyright (c) 2026 Yaku Makki
 * SPDX-License-Identifier: 4P1-Custom
 */

#include <ncurses.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <sys/wait.h>
#include <sys/select.h>
#include <sys/stat.h>
#include <time.h>

#define LINE_MAX 512
#define RING_CAP 20000
#define MAXI(a, b) ((a) > (b) ? (a) : (b))

/* ── line ring (stores ANSI-stripped lines) ── */
static char (*g_ring)[LINE_MAX];
static int g_count;
static int g_follow = 1;
static int g_view;              /* top visible row when not following */

static void strip_ansi(char *s)
{
	char *r = s, *w = s;
	while (*r) {
		if (*r == '\x1b') {
			r++;
			if (*r == '[') {
				r++;
				while (*r && (*r < 0x40 || *r > 0x7e))
					r++;
				if (*r) r++;
			} else if (*r) {
				r++;
			}
		} else {
			*w++ = *r++;
		}
	}
	*w = 0;
}

static void log_push(const char *line)
{
	if (!g_ring) g_ring = malloc(RING_CAP * LINE_MAX);
	if (!g_ring) return;
	if (g_count == RING_CAP) {
		memmove(g_ring, g_ring + 1, (RING_CAP - 1) * LINE_MAX);
		g_count--;
	}
	snprintf(g_ring[g_count], LINE_MAX, "%s", line);
	g_count++;
	if (g_follow)
		g_view = g_count > 0 ? g_count - 1 : 0;
}

/* ── parsing ── */
static int g_pass, g_fail;         /* our tally */
static int g_spass, g_sfail = -1;  /* script's own tally (-1 unknown) */
static char g_section[LINE_MAX];

static void parse_line(const char *s)
{
	if (strncmp(s, "=== ", 4) == 0) {
		snprintf(g_section, LINE_MAX, "%s", s + 4);
		char *e = strstr(g_section, "===");
		if (e) {
			while (e > g_section && (e[-1] == ' ' || e[-1] == '=' ))
				e--;
			*e = 0;
		}
		return;
	}
	if (strncmp(s, "[PASS]", 6) == 0) { g_pass++; return; }
	if (strncmp(s, "[FAIL]", 6) == 0) { g_fail++; return; }
	int a, b;
	if (sscanf(s, "通过: %d", &a) == 1) g_spass = a;
	if (sscanf(s, "失败: %d", &b) == 1) g_sfail = b;
}

/* ── buffered non-blocking line reader over the child pipe ── */
static int g_fd;
static char g_rbuf[8192];
static int g_rlen, g_rpos;
static char g_pend[LINE_MAX];
static int g_pendlen;

/* returns 1: line ready in *out, 0: no data, -1: EOF */
static int read_line(char *out, int outsz)
{
	while (1) {
		while (g_rpos < g_rlen) {
			char ch = g_rbuf[g_rpos++];
			if (ch == '\n') {
				g_pend[g_pendlen] = 0;
				strip_ansi(g_pend);   /* colour codes wrap [PASS] */
				snprintf(out, outsz, "%s", g_pend);
				g_pendlen = 0;
				return 1;
			}
			if (ch != '\r' && g_pendlen < LINE_MAX - 1)
				g_pend[g_pendlen++] = ch;
		}
		fd_set rf;
		struct timeval tv = { 0, 0 };

		FD_ZERO(&rf);
		FD_SET(g_fd, &rf);
		if (select(g_fd + 1, &rf, NULL, NULL, &tv) <= 0)
			return 0;
		ssize_t r = read(g_fd, g_rbuf, sizeof(g_rbuf));
		if (r <= 0)
			return -1;
		g_rlen = (int)r;
		g_rpos = 0;
	}
}

/* ── UI ── */
static double since(clock_t t0)
{
	return (double)(clock() - t0) / CLOCKS_PER_SEC;
}

static void draw(int running, int rc, double elapsed, const char *saved)
{
	erase();
	int my, mx;

	getmaxyx(stdscr, my, mx);
	int ph = my - 3;
	char head[160];

	if (running)
		snprintf(head, sizeof(head),
			 " M4KK1 make test — running %5.1fs", elapsed);
	else
		snprintf(head, sizeof(head),
			 " M4KK1 make test — %s %5.1fs",
			 rc == 0 ? "ALL PASS rc=0" : "FAILURES", elapsed);
	attron(A_BOLD | COLOR_PAIR(running ? 3 : (rc == 0 ? 4 : 1)));
	mvaddnstr(0, 0, head, mx - 1);
	attroff(A_BOLD | COLOR_PAIR(running ? 3 : (rc == 0 ? 4 : 1)));
	char cnt[80];

	snprintf(cnt, sizeof(cnt), "PASS %d  FAIL %d%s", g_pass, g_fail,
		 g_sfail >= 0 ? "  (script: P/F)" : "");
	mvaddnstr(0, MAXI(0, mx - 1 - (int)strlen(cnt)), cnt, mx - 1);
	if (g_section[0])
		mvaddnstr(1, 0, g_section, mx - 1);

	int top = g_follow ? MAXI(0, g_count - ph) : g_view;
	for (int i = 0; i < ph; i++) {
		int idx = top + i;
		if (idx >= g_count)
			break;
		int attr = COLOR_PAIR(2);
		if (strncmp(g_ring[idx], "[PASS]", 6) == 0)
			attr = COLOR_PAIR(3);
		else if (strncmp(g_ring[idx], "[FAIL]", 6) == 0)
			attr = COLOR_PAIR(1) | A_BOLD;
		attron(attr);
		mvaddnstr(2 + i, 0, g_ring[idx], mx - 1);
		attroff(attr);
	}
	char foot[160];

	snprintf(foot, sizeof(foot),
		 " [f]ollow=%s  [s]ave report  [PgUp/PgDn] scroll  [q]uit %s",
		 g_follow ? "on" : "off", saved);
	attron(A_REVERSE);
	mvaddnstr(my - 1, 0, foot, mx - 1);
	attroff(A_REVERSE);
	refresh();
}

static void save_report(const char **outname)
{
	char path[160];
	char ts[64];
	time_t now = time(NULL);
	struct tm *tm = localtime(&now);

	strftime(ts, sizeof(ts), "%Y%m%d_%H%M%S", tm);
	mkdir("logs", 0777);
	snprintf(path, sizeof(path), "logs/tui_test_%s.log", ts);
	FILE *f = fopen(path, "w");
	if (!f)
		return;
	fprintf(f, "# make test report %s  PASS=%d FAIL=%d\n",
		ts, g_pass, g_fail);
	for (int i = 0; i < g_count; i++)
		fprintf(f, "%s\n", g_ring[i]);
	fclose(f);
	*outname = strdup(path + 5);
}

static void handle_key(int k, int ph, const char **saved)
{
	if (k == 'f')
		g_follow = !g_follow;
	else if (k == 's')
		save_report(saved);
	else if (k == KEY_PPAGE)
		{ g_follow = 0; g_view = MAXI(0, g_view - ph); }
	else if (k == KEY_NPAGE)
		{ g_follow = 0; g_view += ph; if (g_view >= g_count) g_view = g_count - 1; }
	else if (k == KEY_UP)
		{ g_follow = 0; g_view = MAXI(0, g_view - 1); }
	else if (k == KEY_DOWN)
		{ g_follow = 0; g_view += 1; if (g_view >= g_count) g_view = g_count - 1; }
}

int main(void)
{
	/* spawn: bash tools/testing/test_all.sh, stdout+stderr -> pipe */
	int pfd[2];
	if (pipe(pfd) != 0) { perror("pipe"); return 2; }
	pid_t child = fork();
	if (child < 0) { perror("fork"); return 2; }
	if (child == 0) {
		close(pfd[0]);
		dup2(pfd[1], STDOUT_FILENO);
		dup2(pfd[1], STDERR_FILENO);
		close(pfd[1]);
		execl("/bin/bash", "bash",
		      getenv("TUI_TEST_SCRIPT")
			  ? getenv("TUI_TEST_SCRIPT")
			  : "tools/testing/test_all.sh",
		      (char *)NULL);
		_exit(127);
	}
	close(pfd[1]);
	g_fd = pfd[0];
	int fl = fcntl(g_fd, F_GETFL);
	fcntl(g_fd, F_SETFL, fl | O_NONBLOCK);

	initscr();
	if (has_colors()) {
		start_color();
		use_default_colors();
		init_pair(1, COLOR_RED, -1);
		init_pair(2, COLOR_WHITE, -1);
		init_pair(3, COLOR_GREEN, -1);
		init_pair(4, COLOR_CYAN, -1);
	}
	curs_set(0);
	keypad(stdscr, TRUE);
	noecho();
	timeout(100);

	const char *saved = "";
	clock_t t0 = clock();
	int rc = -1;
	char line[LINE_MAX];

	while (1) {
		int eof = 0, drained = 0;
		while (drained < 500) {
			int r = read_line(line, sizeof(line));
			if (r == 1) {
				log_push(line);
				parse_line(g_ring[g_count - 1]);
				drained++;
			} else if (r == -1) {
				eof = 1;
				break;
			} else
				break;
		}
		if (eof) {
			int st = 0;
			waitpid(child, &st, 0);
			close(g_fd);
			rc = WIFEXITED(st) ? WEXITSTATUS(st) : 1;
			break;
		}
		int my, mx;

		getmaxyx(stdscr, my, mx);
		(void)mx;
		draw(1, rc, since(t0), saved);
		int k = getch();
		if (k == 'q' || k == 27) {
			endwin();
			kill(child, SIGTERM);
			int st;
			waitpid(child, &st, 0);
			fprintf(stderr, "killed by user\n");
			return 130;
		}
		handle_key(k, my - 3, &saved);
	}

	/* finished: show result until user leaves */
	while (1) {
		int my, mx;

		getmaxyx(stdscr, my, mx);
		(void)mx;
		draw(0, rc, since(t0), saved);
		int k = getch();
		if (k == 'q' || k == 27)
			break;
		if (k == 's')
			save_report(&saved);
		handle_key(k, my - 3, &saved);
	}
	endwin();
	printf("make test: PASS %d  FAIL %d (script tally %d/%d) rc=%d\n",
	       g_pass, g_fail, g_spass, g_sfail, rc);
	return rc;
}
